### ISSUE TITLE
OneCycleLr : which optimizer

### ISSUE THREAD

**Issue by pfeatherstone** [pytorch/pytorch#36133]
Which optimizer is recommended to be used with OneCycleLr? The docs suggest SGD. Is that because the scheduler only has real effect with a bare-bones optimizer?

cc @vincentqb

**Comment by prajjwal1** [comment]
I think you can go with AdamW. The author of the paper used SGD. SGD requires a lot of tuning (in terms of hyperparameters) to work well as compared to Adam counterparts. You can have a look at this [article](https://www.fast.ai/2018/07/02/adam-weight-decay/) for more details. When you use a high learning rate (if trying to acheive superconvergence) be careful with regularization as the author points out in his [paper](https://arxiv.org/pdf/1803.09820.pdf). In short, the answer is you can use other optimizers as well. @gchanan you can close this if the answer seems suitable.

**Comment by pfeatherstone** [comment]
Thank you for the answer. With regards to regularisation, how do you suggest we adapt weight decay when attempting super convergence?

**Comment by prajjwal1** [comment]
When your LR is high (let's say in cyclical LR, max bound is on a higher side), then you'd want to to experiment a bit (running for few epochs on a sample mini dataset) to see how it adapts. There should be a right balance between regularization and right LR in the first half when trying for superconvergence. I'd experiment a few times with `weight_decay` parameter to learn how much regularization would be required.

**Comment by pfeatherstone** [comment]
Do you thing the weight decay can be alo scheduled in OneCycleLr? A bit like momentum, weight decay could decrease when LR is increasing, and increase when LR is decreasing?

**Comment by prajjwal1** [comment]
1CycleLR is a learning rate scheduler. AdamW is an optimizer. So weight decay is incorporated in AdamW not 1CycleLR. Learning rate and momentum increase and decrease at a similar rate simultaneously. We want our optimizer to find a good flat area and during our initial discovery, we want to give more weight to initial gradients. See the figure from this blog [post](https://sgugger.github.io/the-1cycle-policy.html). I hope your queries have been answered.

**Comment by pfeatherstone** [comment]
My point is, the scheduler schedules two hyper parameters of the optimiser, namely: LR and momentum. Why not also schedule a 3rd hyper parameter of the optimiser, namely : the weight decay

**Comment by pfeatherstone** [comment]
> There should be a right balance between regularization and right LR in the first half when trying for superconvergence. I'd experiment a few times with weight_decay parameter to learn how much regularization would be required.
> 
I'm just feeding off what you mentioned above. It seems that weight_decay could benefit from being adapted too.

**Comment by prajjwal1** [comment]
When weight_decay is already present in the [optimizer](https://pytorch.org/docs/stable/optim.html#torch.optim.AdamW), why would you want that in a learning rate scheduler ? This topic is better suited for forums.

**Comment by pfeatherstone** [comment]
Then why would you want the momentum or the beta1 value of the optimizer to be adapted by the scheduler?

**Comment by pfeatherstone** [comment]
According to the docs, the OneCycleLr cycles the momentum between base_momentum and max_momentum

**Comment by pfeatherstone** [comment]
So it's not just a learning rate scheduler, but a hyper-parameter scheduler

**Comment by prajjwal1** [comment]
Yes, it's more than a learning rate scheduler (it's a common terminology), there are other hyper parameters you can adjust with schedulers. I suggest you read the paper. That'll give you more perspective. Optimizer is called on every iteration whereas scheduler is called per epoch. You don't change the learning rate after very forward pass. You learn the heuristics (moving average, warmup etc.) on an epoch and then perform the updation.

**Comment by pfeatherstone** [comment]
> whereas scheduler is called per epoch
> 
The OneCycleLr is called after every batch, not epoch

**Comment by pfeatherstone** [comment]
I think we are going round in circles. I know how and when optimizers and schedulers should be used. My original question has been answered: all optimizers can be used with OneCycleLr in theory. Practically use SGD or AdamW. With AdamW, be careful to use an adequate weight_decay value.

**Comment by pfeatherstone** [comment]
The remainder of the discussion about adapting other hyper parameters in schedulers, I will post on the discuss.pytorch forum at some point.

**Comment by vincentqb** [comment]
Since this discussion is more suited for [the forum](https://discuss.pytorch.org/), I will close this issue. Please feel free to provide the link to it here, and continue the discussion there.
